magenta-lib/src/main/scala/magenta/tasks/SetStackPolicyTask.scala

```diff
@@ -1,82 +1,110 @@
 package magenta.tasks
+import magenta.tasks.StackPolicy.{accountPrivateTypes, sensitiveResourceTypes, toPolicyDoc}
 import magenta.{DeploymentResources, KeyRing, Region}
-import software.amazon.awssdk.services.cloudformation.model.{ChangeSetType, SetStackPolicyRequest}
+import software.amazon.awssdk.services.cloudformation.CloudFormationClient
+import software.amazon.awssdk.services.cloudformation.model.{ChangeSetType, ListTypesRequest, SetStackPolicyRequest, Visibility}
 
-case class StackPolicy(name: String, body: String)
+import scala.collection.JavaConverters.iterableAsScalaIterableConverter
+
+sealed trait StackPolicy { def name: String = this.getClass.getSimpleName }
+case object AllowAllPolicy extends StackPolicy
+case object DenyReplaceDeletePolicy extends StackPolicy
 
 object StackPolicy {
 
-  val ALLOW_ALL_POLICY: StackPolicy = StackPolicy("AllowAll",
-  """{
-    |  "Statement" : [
-    |    {
-    |      "Effect" : "Allow",
-    |      "Action" : "Update:*",
-    |      "Principal": "*",
-    |      "Resource" : "*"
-    |    }
-    |  ]
-    |}
-    |""".stripMargin
-  )
+  def toPolicyDoc(policy: StackPolicy, sensitiveResourceTypes: Set[String], accountPrivateTypes: () => Set[String]): String = policy match {
+    case AllowAllPolicy =>
+      ALLOW_ALL_POLICY
+    case DenyReplaceDeletePolicy =>
+      val sensitiveResources = sensitiveResourceTypes.filter(t => t.startsWith("AWS") || accountPrivateTypes().contains(t))
+      DENY_REPLACE_DELETE_POLICY(sensitiveResources)
+  }
+
+  def accountPrivateTypes(client: CloudFormationClient): Set[String] = {
+    val request =  ListTypesRequest.builder().visibility(Visibility.PRIVATE).build()
+    val response = client.listTypesPaginator(request)
+    response.typeSummaries().asScala.map(_.typeName()).toSet
+  }
+
+  /** CFN resource types that have state or are likely to exist in external
+   * config such as DNS or application config.
+   */
+  val sensitiveResourceTypes: Set[String] =
```

Comment:
minor: I think naming this `allSensitiveResourceTypes` (or similar) would make the tests in `SetStackPolicyTaskTest` a bit easier to read, specifically [this bit](https://github.com/guardian/riff-raff/pull/653/files#diff-340a57d05e31707b6da53a228e9b7c2894479d7b842dbc3bde130de131d8b823R82).